src/Octoshift/Services/MultipartUploaderService.cs

```diff
@@ -0,0 +1,129 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using System.Web;
+using OctoshiftCLI.Extensions;
+
+namespace OctoshiftCLI.Services;
+
+public class MultipartUploaderService
+{
+    private readonly GithubClient _client;
+    internal int _streamSizeLimit = 100 * 1024 * 1024; // 100 MiB
+    internal string _base_url = "https://uploads.github.com/organizations";
+
+    public MultipartUploaderService(GithubClient client)
+    {
+        _client = client;
+    }
+
+    public async Task<string> UploadMultipart(Stream archiveContent, string archiveName, string uploadUrl)
+    {
+        var buffer = new byte[_streamSizeLimit];
+
+        try
+        {
+            // 1. Start the upload
+            var startHeaders = await StartUpload(uploadUrl, archiveName, archiveContent.Length);
+
+            var nextUrl = GetNextUrl(startHeaders);
+            if (nextUrl == null)
+            {
+                throw new OctoshiftCliException("Failed to retrieve the next URL for the upload.");
+            }
+
+            var guid = HttpUtility.ParseQueryString(nextUrl.Query)["guid"];
+
+            // 2. Upload parts
+            int bytesRead;
+            while ((bytesRead = await archiveContent.ReadAsync(buffer, 0, buffer.Length)) > 0)
+            {
+                nextUrl = await UploadPart(buffer, bytesRead, nextUrl.ToString());
+            }
+
+            // 3. Complete the upload
+            await CompleteUpload(nextUrl.ToString());
```

Comment:
also a log here to indicate the upload is done

`Finished uploading the archive.`